### DISCUSSION TITLE
Calling into emscripten native functions

### DISCUSSION THREAD

**Discussion by Beyley** [unoplatform/uno#10179] [Q&A] [answered]
Im trying to call into emscripten OpenGL/GLFW to run GL code on the browser, but im always getting a DLLNotFound error when trying to P/Invoke into __Native or __Internal_emscripten, even with `<WasmShellEmccExportedRuntimeMethod Include="GL" />`, even with the code directly copy pasted from https://github.com/unoplatform/Uno.Wasm.Bootstrap/blob/main/doc/features-module-linking.md#invoking-emscripten-and-mononet-5-native-functions

(this is on wasm bootstrap 3.3.1)

**Comment by jeromelaban** [comment, marked as answer]
Thanks. Can you attach a small repro sample of what you're trying to do?

**Reply by Beyley** [reply]
https://github.com/Beyley/UnoPInvoke
Ideally just call into SDL/OpenGL functions from WASM, but i cant even get to the point of calling into emscripten, as i always get 
```
Error: System.DllNotFoundException: __Internal_emscripten
   at UnoPInvoke.Program.Main(String[] args) in /home/beyley/Projects/PROGRAMMING/C#/UnoPInvoke/UnoPInvoke/Program.cs:line 10
```

**Reply by Beyley** [reply]
my launch script is just `dotnet build && python3 -m http.server -d ./bin/Debug/net6.0/dist/`, then i open it in a browser

**Reply by jeromelaban** [reply]
Thanks for the sample. The documentation may not be clear enough, but unless you specify a `.a` or `.bc` (in which case the bootstrapper automatically picks the name of the file as a candidate for pinvoke), you'll have to explicitly specify the pinvoke name that you will use. 

In your sample, you'll need to add:
```xml
<ItemGroup>
	<WasmShellAdditionalPInvokeLibrary Include="__Internal_emscripten" />
</ItemGroup>
```

**Reply by Beyley** [reply]
> Thanks for the sample. The documentation may not be clear enough, but unless you specify a `.a` or `.bc` (in which case the bootstrapper automatically picks the name of the file as a candidate for pinvoke), you'll have to explicitly specify the pinvoke name that you will use.
> 
> In your sample, you'll need to add:
> 
> ```
> <ItemGroup>
> 	<WasmShellAdditionalPInvokeLibrary Include="__Internal_emscripten" />
> </ItemGroup>
> ```

That makes __Internal_emscripten work, but if i try to do SDL, it fails to compile
```
/home/beyley/.nuget/packages/uno.wasm.bootstrap/3.3.1/build/Uno.Wasm.Bootstrap.targets(175,5): error : System.Exception: Failed to generate AOT layout (More details are available in diagnostics mode or using the MSBuild /bl switch) [/home/beyley/Projects/PROGRAMMING/C#/UnoPInvoke/UnoPInvoke/UnoPInvoke.csproj]
/home/beyley/.nuget/packages/uno.wasm.bootstrap/3.3.1/build/Uno.Wasm.Bootstrap.targets(175,5): error :    at Uno.Wasm.Bootstrap.ShellTask_vfc89b40db92a14b8f04d4c5c9c72690bdcd7b878.RunPackager() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 852 [/home/beyley/Projects/PROGRAMMING/C#/UnoPInvoke/UnoPInvoke/UnoPInvoke.csproj]
/home/beyley/.nuget/packages/uno.wasm.bootstrap/3.3.1/build/Uno.Wasm.Bootstrap.targets(175,5): error :    at Uno.Wasm.Bootstrap.ShellTask_vfc89b40db92a14b8f04d4c5c9c72690bdcd7b878.Execute() in D:\a\1\s\src\Uno.Wasm.Bootstrap\ShellTask.cs:line 228 [/home/beyley/Projects/PROGRAMMING/C#/UnoPInvoke/UnoPInvoke/UnoPInvoke.csproj]
```
I pushed a new commit to the repo showing what im trying for SDL
[binlog](https://github.com/unoplatform/uno/files/9815895/Archive.zip)

**Reply by jeromelaban** [reply]
Got it, thanks. Now you're facing the fact that emscripten does not know that it needs to import SDL, hence this error:

```
[00:00:03.1145008] error: undefined symbol: SDL_CreateWindow (referenced by top-level compiled C/C++ code)
```

You'll need to include [additional emcc flags](https://github.com/unoplatform/Uno.Wasm.Bootstrap/blob/main/doc/features-module-linking.md#static-linking-additional-emcc-flags) (at least):
```xml
<ItemsGroup>
   <WasmShellExtraEmccFlags Include="-s USE_SDL=2" />
</ItemsGroup>
```

for the SDL libs to get linked in.

**Reply by Beyley** [reply]
That seems to have worked! i can call into sdl using manual P/Invoke calls, but im getting a weird error when trying to integrate Silk.NET.SDL into here, specifically `  [00:00:02.9794452] Unhandled exception. System.Exception: Parameter types of pinvoke callback method 'System.Int32 _PTS(Silk.NET.SDL.PfnMallocFunc, Silk.NET.SDL.PfnCallocFunc, Silk.NET.SDL.PfnReallocFunc, Silk.NET.SDL.PfnFreeFunc)' needs to be blittable. (TaskId:111)`
which doesnt make much sense when all the structs listed *should* be blittable
https://github.com/dotnet/Silk.NET/blob/main/src/Windowing/Silk.NET.SDL/Structs/PfnFreeFunc.gen.cs
https://github.com/dotnet/Silk.NET/blob/main/src/Windowing/Silk.NET.SDL/Structs/PfnMallocFunc.gen.cs
https://github.com/dotnet/Silk.NET/blob/main/src/Windowing/Silk.NET.SDL/Structs/PfnCallocFunc.gen.cs
https://github.com/dotnet/Silk.NET/blob/main/src/Windowing/Silk.NET.SDL/Structs/PfnReallocFunc.gen.cs

**Reply by jeromelaban** [reply]
That could be something that is fixed in .NET 7. Can you try with the latest 7.0-dev bootstrapper ?

**Reply by Beyley** [reply]
Im seemingly getting a different error on the latest sdk, but im having trouble understanding the logs
[msbuild.zip](https://github.com/unoplatform/uno/files/9816165/msbuild.zip)

**Reply by jeromelaban** [reply]
That one is a completely different error that's in the bootstrapper itself, and I would not be surprised if it were because of the `#` in your path. Can you try moving the source to another path? Looks like bootstrapper is missing some escaping in that version.

**Reply by Beyley** [reply]
Moved it to a path without a hashtag
[msbuild.log](https://github.com/unoplatform/uno/files/9816213/msbuild.log)

**Reply by Beyley** [reply]
Updated to the latest uno bootstrapper and such and still same issue, updated log incase anything changed:
[msbuild.log](https://github.com/unoplatform/uno/files/9857507/msbuild.log)

**Reply by jeromelaban** [reply]
Thanks. Make sure [to post binlogs](https://platform.uno/docs/articles/uno-builds-troubleshooting.html), not textual log files.

This is very curious, which OS are you using to build?

**Reply by Beyley** [reply]
> Thanks. Make sure [to post binlogs](https://platform.uno/docs/articles/uno-builds-troubleshooting.html), not textual log files.
> 
> This is very curious, which OS are you using to build?

I'm on Linux, specifically Arch Linux, using the net 7 previews

**Reply by jeromelaban** [reply]
I see. Then maybe there's something specific with arch linux and the support for bash, somehow. It looks like some commands don't have the proper structure, somehow. What's your default shell?

**Reply by Beyley** [reply]
> I see. Then maybe there's something specific with arch linux and the support for bash, somehow. It looks like some commands don't have the proper structure, somehow. What's your default shell?

My default shell is zsh, perhaps there's an incompatibility there?

**Reply by jeromelaban** [reply]
If could be, yes. We'll likely have to specify more explicitly which shell needs to be used.

**Reply by Beyley** [reply]
> If could be, yes. We'll likely have to specify more explicitly which shell needs to be used.

i should have bash installed on my machine, is it possible to have the script run under bash directly rather than the users default shell? or is that not possible here

**Reply by jeromelaban** [reply]
The commands are executed with `/bin/sh -c`, so it's probably defaulting to whatever is available. We can make it more explicit to avoid this issue.

**Reply by jeromelaban** [reply]
The Uno.Wasm.Bootstrap package 7.0.0-dev.237 is being published, can you give it a try once you see it?

**Reply by Beyley** [reply]
Still erroring out, heres the proper binlog, with a fake extension at the end because github doesnt let me upload .binlog files
[msbuild.binlog.fake.extension.to.fool.github.just.remove.all.this.and.the.zip.at.the.end.zip](https://github.com/unoplatform/uno/files/9864661/msbuild.binlog.fake.extension.to.fool.github.just.remove.all.this.and.the.zip.at.the.end.zip)

the binlog seems to indicate bash is being used here now

```
┌─[beyley@arch] - [~/Projects/PROGRAMMING/UnoPInvoke/UnoPInvoke] - [2022-10-25 04:41:09]
└─[2] <git:(master 37f436c✱✈) > bash --version
GNU bash, version 5.1.16(1)-release (x86_64-pc-linux-gnu)
Copyright (C) 2020 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is free software; you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
```

**Reply by jeromelaban** [reply]
Thanks for the binlog. It was indeed a regression specific to linux in the 7.0.0 branch in debug mode only. It should be fixed in version 7.0.0-dev.239 of the bootstrapper that will be available in the next 30 minutes.

**Reply by Beyley** [reply]
> Thanks for the binlog. It was indeed a regression specific to linux in the 7.0.0 branch in debug mode only. It should be fixed in version 7.0.0-dev.239 of the bootstrapper that will be available in the next 30 minutes.

That gets it to build fully, but on attempting to run i get a crash
![image](https://user-images.githubusercontent.com/34501060/198093918-b9a99181-f9bc-468b-91a1-674427a74837.png)

**Reply by jeromelaban** [reply]
I see. This one is a fix that still needs to get in, but the workaround is to create a folder named `WasmScripts`, add an empty JS file in it, then make add:
```xml
<ItemGroup>
<EmbeddedResource Include="WasmScripts/**" />
</ItemGroup>
```

This is an issue that is specific to very small projects.

**Reply by Beyley** [reply]
that gets it to launch with all the Silk.NET functions available and compiled properly (though non-functional at the moment, due to us not knowing how to talk with emscripten SDL)

**Reply by jeromelaban** [reply]
Great! Thanks for the update. The next update of the bootstrapper will contain the fix for that last issue you got.
